components/lib/crash/src/main/java/mozilla/components/lib/crash/service/MozillaSocorroService.kt

```diff
@@ -20,54 +39,220 @@ typealias GeckoCrashReporter = CrashReporter
  *                [File a bug](https://bugzilla.mozilla.org/enter_bug.cgi?product=Socorro) if you would like to get your
  *                app added to the whitelist.
  */
+@Suppress("TooManyFunctions", "LargeClass")
 class MozillaSocorroService(
     private val applicationContext: Context,
-    private val appName: String
+    private val appName: String,
+    private val serverUrl: String = defaultServerUrl
 ) : CrashReporterService {
-    override fun report(crash: Crash.UncaughtExceptionCrash) {
-        // Not implemented currently.
+    private val logger = Logger("mozac/MozillaSocorroCrashHelperService")
+    private val startTime = System.currentTimeMillis()
+    val partSet = mutableSetOf<String>()
 
-        // In theory we could upload uncaught exception crashes to Socorro too. But Socorro is not the best tool for
-        // them and it is not used by the app teams.
+    override fun report(crash: Crash.UncaughtExceptionCrash) {
+        try {
+            sendReport(crash.throwable, null, null)
+        } catch (e: IOException) {
+            Logger.error("failed to send uncaught exception crash report to Socorro", e)
+        }
     }
 
     override fun report(crash: Crash.NativeCodeCrash) {
-        // GeckoView comes with a crash reporter class that we can use here. For now we are assuming that we only want
-        // to upload native code crashes to Socorro if GeckoView is used. If this assumption is going to be wrong in
-        // the future then we can start inlining the functionality here.
-        sendViaGeckoViewCrashReporter(crash)
+        try {
+            sendReport(null, crash.minidumpPath, crash.extrasPath)
+        } catch (e: IOException) {
+            Logger.error("failed to send native code crash report to Socorro", e)
+        }
+    }
+
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    internal fun sendReport(throwable: Throwable?, miniDumpFilePath: String?, extrasFilePath: String?) {
+        val serverUrl = URL(serverUrl)
+        val boundary = generateBoundary()
+        partSet.clear()
+
+        with(serverUrl.openConnection() as HttpURLConnection) {
+            requestMethod = "POST"
+            doOutput = true
+            setRequestProperty("Content-Type", "multipart/form-data; boundary=$boundary")
+            setRequestProperty("Content-Encoding", "gzip")
+
+            sendCrashData(outputStream, boundary, throwable, miniDumpFilePath, extrasFilePath)
+
+            BufferedReader(InputStreamReader(inputStream)).use {
+                val response = StringBuffer()
+                var inputLine = it.readLine()
+                while (inputLine != null) {
+                    response.append(inputLine)
+                    inputLine = it.readLine()
+                }
+
+                it.close()
+                Logger.info("Crash reported to Socorro: $response")
+            }
+
+            disconnect()
```

Comment:
I wonder if that needs to go into a `finally` block so that we close it in an error case too.

Reply:
I see.  Missed this case.  I'll see if I can refactor it better.  Thanks,